src/circuits/gates.jl

```diff
@@ -306,6 +309,8 @@ function gate(::GateName"randU", N::Int = 2;
   return Q
 end
 
+gate(::GateName"Haar", N::Int = 2; eltype = ComplexF64, random_matrix = randn(eltype, N, N)) =
```

Comment:
I was also thinking it might be nice to define `"Haar"`/`"randU"` to accept an `rng` directly, to more easily make an individual random gate that you have control over the randomness. Something like:
```julia
function gate(::GateName"randU", N::Int = 2; eltype = ComplexF64, rng = Random.GLOBAL_RNG, random_matrix = randn(rng, eltype, N, N))
  [...]
end
```
This could even make use of the `randomparams` functionality, i.e.:
```julia
function gate(gn::GateName"randU", N::Int = 2; eltype = ComplexF64, rng = Random.GLOBAL_RNG, randomparams(gn, N; eltype = eltype, rng)...)
  [...]
end
```
Not totally necessary, but it could help to create a template like this for users who want to make custom gates that they want to use in a random circuit.

Reply:
I thought I made a comment about the gate name but I can't find it. Can we make this gate name more descriptive? I didn't understand what was meant by a `"Haar"` gate until I saw the definition. Could it be something like `"HaarRandom"`? Also, is a Haar random matrix always unitary (I believe so but maybe you know for sure)? If not, we should be even more specific and use `"HaarRandomUnitary"`. Another name could be `"HaarDistributed"`.

We should add a reference about this algorithm from computing Haar random matrices, for example:  http://math.mit.edu/~edelman/publications/random_matrix_theory.pdf (either in a docstring or comment).

Note that: https://en.wikipedia.org/wiki/Haar_wavelet#Haar_matrix, I doubt people will really confuse it with that definition but still good to be more specific.